R/dm.R

```diff
@@ -732,56 +732,144 @@ as.list.zoomed_dm <- function(x, ...) {
-      cat_line(trim_width(paste0("Primary key: ", pk), glimpse_width))
+#' Print glimpse for a single table included in the `dm` object (zoomed or not)
+#' @keywords internal
+#' @noRd
+print_glimpse_table <- function(x, table_name, width, ...) {
+  if (is_zoomed(x)) {
+    table <- dm_get_zoom(x)$zoom[[1]]
+  } else {
+    table <- x[[table_name]]
+  }
+
+  # `print_glimpse_table_meta()` is not part of this because it needs to be
+  # printed only once for the entire object
+  print_glimpse_table_name(table_name, width)
+  print_glimpse_table_pk(x, table_name, width)
+  print_glimpse_table_fk(x, table_name, width)
+  glimpse(table, width = width, ...)
+}
+
+#' Print table name for a given table in the `dm` object (zoomed or not)
+#' @keywords internal
+#' @noRd
+print_glimpse_table_name <- function(table_name, width) {
+  cat_line("\n", trim_width(paste0("Table: ", tick(table_name)), width))
+}
+
+#' Print details about primary key for a given table in the `dm` object (zoomed or not)
+#' @keywords internal
+#' @noRd
+print_glimpse_table_pk <- function(x, table_name, width) {
+  # return early if there are no primary keys, since there is nothing to print here
+  if (!is_zoomed(x)) {
+    pk_df <- dm_get_all_pks(x)
+
+    if (nrow(pk_df) == 0L) {
+      return(invisible(x))
     }
-    fk <- all_fks %>%
-      filter(child_table == table_name) %>%
-      select(-child_table) %>%
-      pmap_chr(
-        function(child_fk_cols, parent_table, parent_key_cols, on_delete) {
-          trim_width(
-            paste0(
-              "  (",
-              paste0(tick(child_fk_cols), collapse = ", "), ")",
-              " -> ",
-              "(`", paste0(parent_table, "$", parent_key_cols, collapse = "`, `"), "`) ",
-              on_delete
-            ),
-            glimpse_width
-          )
-        }
-      )
-    if (!is_empty(fk)) {
-      cat_line(length(fk), " outgoing foreign key(s):")
-      cat_line(fk)
+  }
+
+  pk <- dm_get_pk_impl(x, table_name)
+
+  # anticipate that some key columns could have been removed by the user
+  if (is_zoomed(x)) {
+    pk <- pk %>%
+      map(~ keep(.x, .x %in% col_tracker_zoomed(x)))
+  }
+
+  pk <- pk %>%
+    map_chr(~ paste0("(", paste0(tick(.x), collapse = ", "), ")"))
+
+  if (!is_empty(pk)) {
+    # FIXME: needs to change if #622 is solved
+    cat_line(trim_width(paste0("Primary key: ", pk), width))
+  }
+}
+
+#' Print details about foreign keys for a given table in the `dm` object (zoomed or not)
+#' @keywords internal
+#' @noRd
+print_glimpse_table_fk <- function(x, table_name, width) {
+  # return early if there are no foreign keys, since there is nothing to print here
+  if (!is_zoomed(x)) {
+    fk_df <- dm_get_all_fks(x)
+
+    if (nrow(fk_df) == 0L) {
+      return(invisible(x))
     }
-    glimpse(table, width = width, ...)
-  })
-  invisible(x)
+  }
+
+  all_fks <- dm_get_all_fks_impl(x)
+
+  fk <- all_fks %>%
+    filter(child_table == table_name) %>%
```

Comment:
I try to be explicit, even if we control this data structure. No need to act now.

```suggestion
    filter(child_table == !!table_name) %>%
```